### PR TITLE
feat(qwen3): device-side greedy sampling and embedding

### PR DESCRIPTION
**Pull request by sunkaixuan2018**
## Summary

This PR adds the Qwen3 pypto-lib side of moving greedy sampling and token embedding onto device.

It is paired with the pypto-serving PR below. The two PRs should be reviewed together: pypto-lib provides the device kernels and padded-vocab semantics, while pypto-serving wires those kernels into generation and validates the serving behavior.

Paired pypto-serving PR: https://github.com/hw-native-sys/pypto-serving/pull/47

## Performance Impact

Measured with the paired serving PR on myserver using Qwen3-14B, batch=1, short prompt `Huawei`, greedy temperature=0, `max_new_tokens=128`, full 40-layer NPU generation:

- Latest main baseline: total 8.583s, prefill 0.410s, decode 6.211s over 127 decode steps, 48.9 ms/token.
- Device emb/sample branch: total 6.450s, prefill 0.373s, decode 6.033s over 127 decode steps, 47.5 ms/token.

This is about 25% lower end-to-end generation time in that run, with decode TPOT slightly lower as well. The main reason is not a large kernel-level speedup: the device path adds small `greedy_sample` / `token_embed` kernels, but removes the per-token host sampling, host embedding lookup, and host/device synchronization work from the generation loop.

## What Changed

- Add `REAL_VOCAB` for Qwen3-14B so kernels can distinguish the real vocabulary from the padded device vocabulary.
- Add standalone greedy sampling and token embedding kernels.
- Integrate previous-token embedding and next-token greedy sampling into the fused decode kernel.
- Keep prefill free of inline sample/embed logic; serving runs standalone device greedy sampling after prefill when greedy generation is enabled.
- Make greedy tie-break match host `torch.argmax` by reverse-scanning equal best logits so the smallest token id wins.
- Clamp any sampled padded-vocab id back to token 0 as a defensive guard.

## Validation

Validated together with the paired serving branch on myserver:

- `pytest tests/test_batching.py tests/test_device_sampling_submission.py -q` -> 20 passed
- `python -m pytest tests/golden/test_qwen3_greedy_source.py -q` -> 3 passed
- greedy sample `a2a3sim` -> PASS
- NPU generation -> exit 0, token ids `[594, 220, 20, 38, 5440]`, text `'s 5G technology`

### CODE REVIEW

File: models/qwen3/14b/decode_layer.py [hunk 1]
```diff
@@ -1104,6 +1118,75 @@ def _decode_layer(  # noqa: PLR0913 — model signature is intrinsic
...
+
+            local_token = pl.cast(0, pl.INT32)
+            chunk_base = chunk_i32 * pl.cast(SAMPLE_VOCAB_CHUNK, target_type=pl.INT32)
+            for t in pl.range(SAMPLE_VOCAB_CHUNK):
+                scan_t = (SAMPLE_VOCAB_CHUNK - 1) - t
+                token_i32 = chunk_base + pl.cast(scan_t, target_type=pl.INT32)
+                token_idx = pl.cast(token_i32, target_type=pl.INDEX)
+                val = pl.read(logits, [b, token_idx])
+                if val == best_val:
+                    local_token = pl.cast(scan_t, pl.INT32)
```
[the remark this answers is not in the document]
Comment (sunkaixuan2018) [reply models/qwen3/14b/decode_layer.py]: Fixed in 805d55b - decode_layer now slices the winning logits chunk once into a local tensor before scanning for the tie-break token.

File: models/qwen3/14b/greedy_sample.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+            local_token = pl.cast(0, pl.INT32)
+            chunk_base = chunk_i32 * pl.cast(VOCAB_CHUNK, target_type=pl.INT32)
+            for t in pl.range(VOCAB_CHUNK):
+                scan_t = (VOCAB_CHUNK - 1) - t
+                token_i32 = chunk_base + pl.cast(scan_t, target_type=pl.INT32)
+                token_idx = pl.cast(token_i32, target_type=pl.INDEX)
+                val = pl.read(logits, [b, token_idx])
+                if val == best_val:
+                    local_token = pl.cast(scan_t, pl.INT32)
```
[the remark this answers is not in the document]
Comment (sunkaixuan2018) [reply models/qwen3/14b/greedy_sample.py]: Fixed in 805d55b - greedy_sample now reads the winning chunk from a local slice instead of repeatedly reading global logits in the sequential scan.

File: models/qwen3/14b/prefill_fwd.py [hunk 3]
```diff
@@ -560,6 +575,77 @@ def prefill_layer(
...
+
+            local_token = pl.cast(0, pl.INT32)
+            chunk_base = chunk_i32 * pl.cast(SAMPLE_VOCAB_CHUNK, target_type=pl.INT32)
+            for t in pl.range(SAMPLE_VOCAB_CHUNK):
+                scan_t = (SAMPLE_VOCAB_CHUNK - 1) - t
+                token_i32 = chunk_base + pl.cast(scan_t, target_type=pl.INT32)
+                token_idx = pl.cast(token_i32, target_type=pl.INDEX)
+                val = pl.read(logits, [b, token_idx])
+                if val == best_val:
+                    local_token = pl.cast(scan_t, pl.INT32)
```
[the remark this answers is not in the document]
Comment (sunkaixuan2018) [reply models/qwen3/14b/prefill_fwd.py]: Fixed in 805d55b - prefill_fwd now uses a local slice for the winning logits chunk before the token tie-break scan.

File: models/qwen3/14b/decode_layer.py [hunk 4]
```diff
@@ -1104,6 +1118,75 @@ def _decode_layer(  # noqa: PLR0913 — model signature is intrinsic
...
+                hidden_chunk = pl.slice(embed_weight, [1, EMBED_HIDDEN_CHUNK], [token_row, k0])
+                next_hidden = pl.assemble(next_hidden, hidden_chunk, [b, k0])
+    return next_hidden
+
+
+@pl.jit.inline
+def _greedy_sample_inline(
+    logits: pl.Tensor[[BATCH, VOCAB], pl.FP32],
+    sampled_ids: pl.Tensor[[BATCH, SAMPLED_IDS_PAD], pl.INT32],
+) -> pl.Tensor[[BATCH, SAMPLED_IDS_PAD], pl.INT32]:
```
[the remark this answers is not in the document]
Comment (sunkaixuan2018) [reply models/qwen3/14b/decode_layer.py:1142]: Addressed in 805d55b for the duplicated performance issue. I kept the three inline bodies separate because they have different PyPTO JIT signatures and compile-time constants: standalone greedy sampling uses fixed BATCH, prefill uses dynamic user_batch, and decode is fused with embedding. Extracting a shared helper would add JIT-boundary risk in this PR.

File: models/qwen3/14b/decode_layer.py [hunk 5]
```diff
@@ -1132,13 +1215,17 @@ def decode_fwd(  # noqa: PLR0913 — device-side fused NUM_LAYERS decode + LM he
...
-    # head. Weights are STACKED [_FWD_NLAYERS*HIDDEN, ...] / [_FWD_NLAYERS*INTERMEDIATE,
-    # ...]; k_cache / v_cache cover [_FWD_NLAYERS*BATCH*NUM_KV_HEADS*MAX_SEQ, ...]; out
-    # is logits [BATCH, VOCAB]. _FWD_NLAYERS defaults to NUM_LAYERS (40) and is settable
-    # for layer-count tests.
+    # Device-side fused decode: embed the previous sampled token id, loop the inline
+    # body over all _FWD_NLAYERS layers, run the LM head, then sample the next token
+    # id. Weights are STACKED [_FWD_NLAYERS*HIDDEN, ...] /
+    # [_FWD_NLAYERS*INTERMEDIATE, ...]; k_cache / v_cache cover
+    # [_FWD_NLAYERS*BATCH*NUM_KV_HEADS*MAX_SEQ, ...]; out is logits [BATCH, VOCAB].
+    # _FWD_NLAYERS defaults to NUM_LAYERS (40) and is settable for layer-count tests.
```
[the remark this answers is not in the document]
Comment (sunkaixuan2018) [reply models/qwen3/14b/decode_layer.py:1253]: Fixed in 805d55b - the decode comment now correctly says cur is seeded from next_hidden after embedding the previous sampled token id.

File: models/qwen3/14b/prefill_fwd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunkaixuan2018) [reply models/qwen3/14b/prefill_fwd.py]: Follow-up in a2dd19c: I kept decode_layer and standalone greedy_sample on the local chunk-slice optimization, but restored prefill_fwd to the direct dynamic-batch token scan after the a2a3 run hit an AICore runtime failure on the sliced form. This keeps the CI-stable prefill path while preserving the fixed-batch performance fix where it passed device validation.

[the remark this answers is not in the document]
Comment (sunkaixuan2018) [reply models/qwen3/14b/prefill_fwd.py]: Follow-up in 20f60c9: I removed the fused prefill sampling/embedding path after both prefill_fwd.py and qwen3_14b_prefill_tq.py continued to fail on a2a3 with runtime 507018. Prefill is back to logits-only, while the standalone greedy_sample.py and token_embed.py kernels remain available for serving to compose after prefill. This keeps the device-validated kernel boundary clean.

File: models/qwen3/14b/decode_layer.py [hunk 6]
```diff
@@ -1104,6 +1118,75 @@ def _decode_layer(  # noqa: PLR0913 — model signature is intrinsic
...
+            chunk_base_idx = pl.cast(chunk_base, target_type=pl.INDEX)
+            winning_logits = pl.slice(logits, [1, SAMPLE_VOCAB_CHUNK], [pl.cast(b, pl.INDEX), chunk_base_idx])
+            for t in pl.range(SAMPLE_VOCAB_CHUNK):
+                scan_t = (SAMPLE_VOCAB_CHUNK - 1) - t
+                val = pl.read(winning_logits, [0, pl.cast(scan_t, pl.INDEX)])
+                if val == best_val:
+                    local_token = pl.cast(scan_t, pl.INT32)
+            token_id = chunk_base + local_token
+            if token_id >= pl.cast(REAL_VOCAB, target_type=pl.INT32):
+                token_id = pl.cast(0, pl.INT32)
```
[the remark this answers is not in the document]
Comment (sunkaixuan2018) [reply models/qwen3/14b/decode_layer.py]: Fixed in db3bf3f - decode greedy sampling now iterates only real-vocab chunks and masks the final partial chunk with fillpad(min), so padded logits cannot win and then clamp to token 0.

File: models/qwen3/14b/greedy_sample.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    def init_logits():
+        logits = torch.full((BATCH, VOCAB), -1000.0)
+        logits[0::2, 0] = 0.0
+        logits[0::2, 7] = 5.0
+        logits[0::2, 42] = 5.0
+        logits[1::2, 0] = 5.0
+        logits[1::2, 7] = 0.0
+        logits[:, REAL_VOCAB:] = logits[:, :1]
+        return logits
```
[the remark this answers is not in the document]
Comment (sunkaixuan2018) [reply models/qwen3/14b/greedy_sample.py:126]: Fixed in db3bf3f - standalone greedy_sample now uses the same real-vocab chunk bound and tail mask as decode. I also added a padded-tail-high fixture so the golden keeps validating real-vocab-only sampling.
